csrc/device_lower/lower2device.cpp

```diff
@@ -206,14 +206,30 @@ void segmenterHintCleanup(Fusion* fusion) {
   }
 }
 
+std::tuple<Val*, Val*, GetRNGSeedAndOffsetFromHost*>
+getRNGSeedAndOffsetFromHost();
+
 void assignRNGOffset(Fusion* fusion) {
-  int counter = 0;
+  Val* seed = nullptr;
+  Val* first_offset = nullptr;
+  GetRNGSeedAndOffsetFromHost* getseed_op = nullptr;
+  int64_t counter = 0;
   for (auto expr : fusion->exprs()) {
-    if (expr->isA<RNGOp>()) {
-      auto rop = expr->as<RNGOp>();
-      rop->setRNGOffset(counter++);
+    if (auto rop = dynamic_cast<RNGOp*>(expr)) {
+      if (!rop->isDeterministic()) {
+        if (seed == nullptr) {
+          std::tie(seed, first_offset, getseed_op) =
+              getRNGSeedAndOffsetFromHost();
+        }
+        Val* offset = SimplifyingIrBuilder::addExpr(first_offset, counter);
+        rop->setSeedAndOffset(seed, offset);
```

Comment:
I was originally thinking of doing it at definition time, but I only came up with an approach that each `RNGOp` gets its own `GetRNGSeedAndOffsetFromHost`, so I disliked it. With the fusion managed  data, it makes much more sense now, but still there are other problems. Because, in the fusion definition, you will have something like
```CUDA
seed = (seed_ptr == nullptr ? *seed_ptr : seed_val)
```
When turning into code, expr sorting will give you something like
```
bool b0 = seed_ptr == nullptr;
int64_t i0 = *seed_ptr;
int64_t seed = b0 ? i0 :  seed_val;
```
This totally makes no sense, because we are dereferencing the pointer regardless of if it is `nullptr` or not. Due to this limitation, `GetRNGSeedAndOffsetFromHost` must be created in the kernel and it must be after expr sorting.